### PR TITLE
Adds Bellman ford algorithm and its test cases.

### PR DESCRIPTION
**Pull request by Kpraful**
This PR adds BellmanFord algorithm and its test cases following the guideline and references from previous code.
- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Comments in areas I changed are up to date
- [x] I have added comments to hard-to-understand areas of my code
- [ ] I have made corresponding changes to the README.md

### CODE REVIEW

**Review by siriak — CHANGES_REQUESTED** [review]
Could you decompose the algorithm into smaller functions so that it's easier to read and maintain?

**Comment by Kpraful** [comment]
Hi, Can you please review PR.

**Review by Kpraful — COMMENTED** [review]
Followed the requested review and break it into smaller functions

**Review by Kpraful — COMMENTED** [review]
Reviewed changes

File: Algorithms/Graph/BellmanFord.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            RelaxEdges();
+            CheckForNegativeCycles();
+            return distances;
+        }
+
+        private void InitializeDistances(Vertex<T> sourceVertex)
+        {
+            foreach (var vertex in graph.Vertices)
+            {
+                if (vertex != null)
```
Comment (siriak) [inline Algorithms/Graph/BellmanFord.cs:44]: Can it be null?

File: Algorithms/Graph/BellmanFord.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        /// <returns>
+        /// A dictionary containing the shortest distances from the source vertex to all other vertices.
+        /// If a vertex is unreachable from the source, it will have a value of double.PositiveInfinity.
+        /// </returns>
+        public Dictionary<Vertex<T>, double> Run(Vertex<T> sourceVertex)
+        {
+            InitializeDistances(sourceVertex);
+            RelaxEdges();
+            CheckForNegativeCycles();
+            return distances;
```
Comment (siriak) [inline Algorithms/Graph/BellmanFord.cs:37]: Please make
```
        private readonly DirectedWeightedGraph<T> graph;
        private readonly Dictionary<Vertex<T>, double> distances;
        private readonly Dictionary<Vertex<T>, Vertex<T>?> predecessors;
```
method parameters, then you won't have to worry about shared state when calling the `Run` method multiple times

File: Algorithms/Graph/BellmanFord.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                    }
+                }
+            }
+        }
+
+        private void RelaxEdgesForVertex(Vertex<T> u)
+        {
+            foreach (var neighbor in graph.GetNeighbors(u))
+            {
+                if (neighbor == null)
```
Comment (siriak) [inline Algorithms/Graph/BellmanFord.cs:74]: The same question, can it be null? It seems odd to me to return null from `GetNeighbors`

**Review by siriak — CHANGES_REQUESTED** [review]

File: Algorithms/Graph/BellmanFord.cs (hunk 3 above)
Comment (siriak) [reply Algorithms/Graph/BellmanFord.cs:74]: I have just checked the implementation of the graph class and it's odd, but your check seems valid

File: Algorithms/Graph/BellmanFord.cs (hunk 1 above)
Comment (siriak) [reply Algorithms/Graph/BellmanFord.cs:44]: I have just checked the implementation of the graph class and it's odd, but your check seems valid

File: Algorithms/Graph/BellmanFord.cs (hunk 2 above)
Comment (Kpraful) [reply Algorithms/Graph/BellmanFord.cs:37]: Hi, I have made the changes. Hence resolving the conversation.

**Review by Kpraful — COMMENTED** [review]
Made changes as adviced.

**Review by siriak — APPROVED** [review]
Looks good, thanks!

**Pull request merged**